components/match2/commons/opponent.lua

```diff
@@ -0,0 +1,331 @@
+StarcraftMatchSummary
+- Wikis may add additional wiki-specific fields to the opponent representation.
+
+]]
+local Opponent = {}
+
+Opponent.partyTypes = {'solo', 'duo', 'trio', 'quad'}
+Opponent.types = Array.extend(Opponent.partyTypes, {'team', 'literal'})
+
+Opponent.partySizes = {
+	solo = 1,
+	duo = 2,
+	trio = 3,
+	quad = 4,
+}
+
+function Opponent.typeIsParty(type)
+	return Opponent.partySizes[type] ~= nil
+end
+
+--[[
+Returns the player count for a party type, or nil otherwise.
+
+Opponent.partySize('duo') == 2
+]]
+function Opponent.partySize(type)
+	return Opponent.partySizes[type]
+end
+
+--[[
+Creates a blank opponent of the specified type
+]]
+function Opponent.blank(type)
+	if type == 'team' then
+		return {type = type, template = 'tbd'}
+	elseif type == 'literal' then
+		return {type = type, name = ''}
+	else
+		assert(Opponent.typeIsParty(type))
+		return {
+			type = type,
+			players = Array.map(
+				Array.range(1, Opponent.partySize(type)),
+				function(_) return {displayName = ''} end
+			),
+		}
+	end
+end
+
+--[[
+Creates a TBD opponent of the specified type
+]]
+function Opponent.tbd(type)
+	if type == 'team' then
+		return {type = type, template = 'tbd'}
+	elseif type == 'literal' then
+		return {type = type, name = 'TBD'}
+	else
+		assert(Opponent.typeIsParty(type))
+		return {
+			type = type,
+			players = Array.map(
+				Array.range(1, Opponent.partySize(type)),
+				function(_) return {displayName = 'TBD'} end
+			),
+		}
+	end
+end
+
+--[[
+Whether an opponent is TBD
+]]
+function Opponent.isTbd(opponent)
+	if opponent.type == 'team' then
+		return opponent.template == 'tbd'
+
+			-- The following can't occur in valid opponents, but we check for them anyway
+			or opponent.name == 'TBD'
+			or String.isEmpty(opponent.template)
+
+	elseif opponent.type == 'literal' then
+		return true
+
+	else
+		return Array.any(opponent.players, Opponent.playerIsTbd)
+	end
+end
+
+function Opponent.playerIsTbd(player)
+	return player.displayName == '' or player.displayName == 'TBD'
+end
+
+--[[
+Whether an arbitary value is a valid representation of an opponent
+]]
+function Opponent.isOpponent(opponent)
+	if not (type(opponent) == 'table' and type(opponent.type) == 'string') then
+		return false
+	end
+
+	if opponent.type == 'team' then
+		return type(opponent.template) == 'string'
+			and opponent.template ~= ''
+
+	elseif opponent.type == 'literal' then
+		return type(opponent.name) == 'string'
+
+	elseif Opponent.typeIsParty(opponent.type) then
+		return type(opponent.players) == 'table'
+			and #opponent.players == Opponent.partySize(opponent.type)
+			and Array.all(opponent.players, Opponent.isPlayer)
+
+	else
+		return false
+	end
+end
+
+function Opponent.isPlayer(player)
+	return type(player) == 'table'
+		and type(player.displayName) == 'string'
+		and (type(player.pageName) == 'string' or player.pageName == nil)
+		and (type(player.flag) == 'string' or player.flag == nil)
+end
+
+function Opponent.isType(type)
```

Comment:
If we have this function why do we still need Opponent.isParty, etc?

Reply:
Opponent.typeIsParty identifies the party opponent type (solo, duo, trio, quad). This identifies all opponent types